### PR TITLE
[libimobiledevice/*] Use original upstream

### PR DESCRIPTION
**Pull request by xiaozhuai**
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] SHA512s are updated for each updated download
- [x] The "supports" clause reflects platforms that may be fixed by this new version
- [x] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [x] Any patches that are no longer applied are deleted from the port's directory.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

---

There are more work to be done: 
* [x] `getopt-win32` use original upstream #33893
* [x] Add `imobiledevice` feature for `kf5solid` #33888
* [x] Add tools feature for `libplist` #33837 
* [x] Add tools feature for `libusbmuxd` #33838
* [x] Add tools feature for `libimobiledevice` #33872
* [x] Add tools feature for `libirecovery` #33868
* [x] Add tools feature for `libideviceactivation` #33871
* [x] ~Patch `idevicerestore` to support windows~ **Not Planned**
* [x] Patch `ideviceinstaller` to support windows
* [x] ~Patch `usbmuxd` to support windows~ **Not Planned**

Since this pr is too big enough, I decide to open other PRs to solve them once this pr get merged.

---

~These lines are added into ci.baseline.txt~
```
idevicerestore:arm64-windows=fail
idevicerestore:x64-windows=fail
idevicerestore:x64-windows-static=fail
idevicerestore:x64-windows-static-md=fail
idevicerestore:x86-windows=fail
ideviceinstaller:arm64-windows=fail
ideviceinstaller:x64-windows=fail
ideviceinstaller:x64-windows-static=fail
ideviceinstaller:x64-windows-static-md=fail
ideviceinstaller:x86-windows=fail
usbmuxd:arm64-windows=fail
usbmuxd:x64-windows=fail
usbmuxd:x64-windows-static=fail
usbmuxd:x64-windows-static-md=fail
usbmuxd:x86-windows=fail
```

~`idevicerestore`, `ideviceinstaller` and `usbmuxd` are tool only port.~
~Before this pr, they only supports windows.~
~But as the upstream said, it should support other platforms.~
~Now it supports `!uwp & !android & !ios & !xbox`, but more patch needed to get windows work, and I would like to leave this job to others who are interested. This is why I add these lines into ci baseline.~

---

Also tested the following triplet:

1. x64-osx-dynamic
2. x64-linux-dynamic

### CODE REVIEW

**Comment by xiaozhuai** [comment]
@Adela0814 @LilyWangL @dg0yt 
I'm able to patch original upstream and build these libs on windows.
The upstream's dllexport must be patched to build dynamic library, that makes the patch very big, as you can see.
What's your opinion if I make these libs static only, so I can make the patches much smaller?

**Comment by dg0yt** [comment]
I appreciate these port being changed to use the original upstream.

You shouldn't move the CMakeLists.txt into patches. It make things more difficult.

> And the win32 fork's version is great than the original, which make we can not use the originial upstrem's version. That sucks.

True, this was one of the show-stoppers for me when I looked at these ports in the past.
I wish I had a good answer.
(But I really see this as a grave port bug: Choosing an inofficial fork leading to wrong versioning w.r.t official upstream.)

> The upstream's dllexport must be patched to build dynamic library, that makes the patch very big, as you can see.

Did you consider using a DEF file instead?
(My most recent approach to semi-automated generation of DEF files for maintainers is in port minizip, cf. https://github.com/microsoft/vcpkg/pull/33044/commits/9321697016b8c88fb63c2281a45f39881bba2a63. After all, I have to do this in vcpkg CI, just accessing the failure logs.)

**Comment by xiaozhuai** [comment]
> Did you consider using a DEF file instead?

@dg0yt 
Looks like indeed what I need. I'll give it a try.

---

BTW, there are some tool only ports, `idevicerestore`, `ideviceinstaller`, `usbmuxd`. Should I fix them or remove them?
As far as I know, vcpkg does not accept tool only port.

**Comment by dg0yt** [comment]
> BTW, there are some tool only ports, `idevicerestore`, `ideviceinstaller`, `usbmuxd`. Should I fix them or remove them?
> As far as I know, vcpkg does not accept tool only port.

Removal is more sensitive than addition.
You may propose this in another PR (unless it becomes a barrier), or annotate the initial post.
For now, they are probably useful as an integration test in vcpkg ci (if not skipped).

**Comment by xiaozhuai** [comment]
@Adela0814 @LilyWangL @dg0yt 

Finally, the ship set sail.

There are more work to be done: 
* [ ] Add `imobiledevice` feature for `kf5solid`.
* [ ] Add tools feature for `libplist`
* [ ] Add tools feature for `libusbmuxd`
* [ ] Add tools feature for `libimobiledevice`
* [ ] Add tools feature for `libirecovery`
* [ ] Add tools feature for `libideviceactivation`

Since this pr is big enough, I decide to open other PRs once this pr is merged.

**Comment by xiaozhuai** [comment]
Also tested the following triplet:
1. x64-osx-dynamic
2. x64-linux-dynamic

**Comment by xiaozhuai** [comment]
The following usage passed on `x64-windows` and `x64-osx`

```cmake
cmake_minimum_required(VERSION 3.15)

project(vcpkg_demo_imobiledevice)

set(CMAKE_CXX_STANDARD 14)
set(CMAKE_CXX_STANDARD_REQUIRED ON)

find_package(unofficial-libplist CONFIG REQUIRED)
find_package(unofficial-libimobiledevice CONFIG REQUIRED)

file(GLOB_RECURSE PROJECT_SOURCES src/*.cpp src/*.hpp src/*.c src/*.h)
add_executable(${PROJECT_NAME} ${PROJECT_SOURCES})
target_link_libraries(${PROJECT_NAME} PRIVATE
        unofficial::libplist::libplist
        unofficial::libplist::libplist++
        unofficial::libimobiledevice::libimobiledevice
)
```

```cpp
/**
 * Copyright 2022 xiaozhuai
 */

#include <cstdio>
#include <vector>

#include "libimobiledevice/libimobiledevice.h"
#include "libimobiledevice/lockdown.h"

#define TOOL_NAME "ideviceinfo"

int main() {
    idevice_info_t *devices = nullptr;
    int device_count = 0;
    idevice_get_device_list_extended(&devices, &device_count);
    for (int i = 0; i < device_count; i++) {
        if (devices[i]->conn_type != CONNECTION_USBMUXD) {
            continue;
        }
        idevice_error_t ret = IDEVICE_E_UNKNOWN_ERROR;
        lockdownd_error_t ld_ret = LOCKDOWN_E_UNKNOWN_ERROR;
        idevice_t device = nullptr;
        lockdownd_client_t client = nullptr;

        ret = idevice_new_with_options(&device, devices[i]->udid, IDEVICE_LOOKUP_USBMUX);
        if (ret != IDEVICE_E_SUCCESS) {
            printf("Error: Failed to connect to device %s, %d\n", devices[i]->udid, ret);
            continue;
        }

        ld_ret = lockdownd_client_new_with_handshake(device, &client, TOOL_NAME);
        if (ld_ret != LOCKDOWN_E_SUCCESS) {
            printf("Error: Failed to connect to lockdownd: %s (%d)\n", lockdownd_strerror(ld_ret), ld_ret);
            idevice_free(device);
            continue;
        }

        char *device_name = nullptr;
        lockdownd_get_device_name(client, &device_name);

        printf("- [%d]\n", i);
        printf("  UDID      : %s\n", devices[i]->udid);
        printf("  Conn Type : %s\n", devices[i]->conn_type == CONNECTION_USBMUXD ? "USB" : "Network");
        printf("  Name      : %s\n", device_name);

        free(device_name);
        lockdownd_client_free(client);
        idevice_free(device);
    }

    return 0;
}
```

**Comment by xiaozhuai** [comment]
@Adela0814 
This PR is ready to review~

File: ports/ideviceinstaller/portfile.cmake [hunk 1]
```diff
@@ -1,19 +1,22 @@
...
-    ADDITIONAL_LIBS getopt.lib
-)
-
-vcpkg_install_copyright(FILE_LIST "${SOURCE_PATH}/COPYING")
-
-set(VCPKG_POLICY_EMPTY_INCLUDE_FOLDER enabled)
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO libimobiledevice/ideviceinstaller
+    REF b9cfe0b264f66eab9ad88e11eb6b0523cb1de911 # commits on 2023-07-21
```
Comment (Adela0814) [inline ports/ideviceinstaller/portfile.cmake:4]: Please restore the EOL of all files below.

**Review by Adela0814 — CHANGES_REQUESTED** [review]

File: scripts/ci.baseline.txt [hunk 2]
```diff
@@ -527,6 +527,21 @@ intelrdfpmathlib:arm64-android=fail
 intelrdfpmathlib:x64-android=fail
 irrlicht:x64-android=fail
 isal:x64-android=fail
+idevicerestore:arm64-windows=fail
+idevicerestore:x64-windows=fail
+idevicerestore:x64-windows-static=fail
```
[the remark this answers is not in the document]
Comment (xiaozhuai) [reply scripts/ci.baseline.txt:532]: See https://github.com/microsoft/vcpkg/pull/33246#issue-1856044076

File: ports/ideviceinstaller/vcpkg.json [hunk 3]
```diff
@@ -1,21 +1,21 @@
 {
   "name": "ideviceinstaller",
-  "version": "1.1.2.23",
-  "port-version": 7,
+  "version-date": "2023-07-21",
```
[the remark this answers is not in the document]
Comment (xiaozhuai) [reply ports/ideviceinstaller/vcpkg.json:3]: I don't think the upstream will do so. We can use `version-date` for now, Once the upstream version number is greater than the win32 fork, we can switch back to `version`

File: ports/libideviceactivation/002_fix_api.patch [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+@@ -32,15 +32,11 @@
+ #include <libxml/HTMLtree.h>
+ #include <curl/curl.h>
+ 
+-#ifdef WIN32
+-#define IDEVICE_ACTIVATION_API __declspec( dllexport )
+-#else
+-#ifdef HAVE_FVISIBILITY
++#if !defined(_WIN32) && !defined(LIBIDEVICEACTIVATION_STATIC)
+ #define IDEVICE_ACTIVATION_API __attribute__((visibility("default")))
```
[the remark this answers is not in the document]
Comment (xiaozhuai) [reply ports/libideviceactivation/002_fix_api.patch:14]: It's just fix api export for windows, why we need to wait for upstream approval?

File: ports/libimobiledevice-glue/004_fix_api.patch [hunk 5]
```diff
@@ -0,0 +1,20 @@
+diff --git a/src/common.h b/src/common.h
+index bd22e3d..144a799 100644
+--- a/src/common.h
++++ b/src/common.h
+@@ -25,14 +25,10 @@
+ #include <config.h>
+ #endif
+ 
+-#ifdef WIN32
```
[the remark this answers is not in the document]
Comment (xiaozhuai) [reply ports/libimobiledevice-glue/004_fix_api.patch:9]: No, If I'm right, and upstream would like to accept these changes, then the `win32` fork wouldn't exists.

[the remark this answers is not in the document]
Comment (xiaozhuai) [reply ports/libimobiledevice-glue/004_fix_api.patch:9]: And there are so many patches, submit all of these to upstream and wait for approval. I don't want to do this : (

File: ports/libimobiledevice/exports.def [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+
+    ; src/syslog_relay.c
+    syslog_relay_client_new
+    syslog_relay_client_start_service
+    syslog_relay_client_free
+    syslog_relay_receive
+    syslog_relay_receive_with_timeout
+    syslog_relay_start_capture
+    syslog_relay_start_capture_raw
+    syslog_relay_stop_capture
```
[the remark this answers is not in the document]
Comment (xiaozhuai) [reply ports/libimobiledevice/exports.def]: Done

File: ports/libimobiledevice-glue/004_fix_api.patch (hunk 5 above)
[the remark this answers is not in the document]
Comment (xiaozhuai) [reply ports/libimobiledevice-glue/004_fix_api.patch:9]: BTW, some of the patches here are not best practices, in order to keep them as small as possible. If submit a pr to upstream, then all of the patches need to update.
See https://github.com/microsoft/vcpkg/pull/33246#issuecomment-1687427346 and https://github.com/microsoft/vcpkg/pull/33246#issuecomment-1687452019

File: ports/ideviceinstaller/portfile.cmake (hunk 1 above)
Comment (xiaozhuai) [reply ports/ideviceinstaller/portfile.cmake:4]: Considering that these portfile.cmake are almost rewritten, I think is the best time to change the newline character.

Comment (xiaozhuai) [reply ports/ideviceinstaller/portfile.cmake:4]: After thinking about it for a while, let's change it back.

Comment (xiaozhuai) [reply ports/ideviceinstaller/portfile.cmake]: Done

File: ports/libimobiledevice-glue/004_fix_api.patch (hunk 5 above)
[the remark this answers is not in the document]
Comment (Adela0814) [reply ports/libimobiledevice-glue/004_fix_api.patch:9]: I know it's difficult, but this is our strategy. If a patch could possibly be useful by upstream, upstream must be notified of the patch's content.
see https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md#:~:text=If%20a%20patch%20could%20possibly%20be%20useful%20by%20upstream%2C%20upstream%20must%20be%20notified%20of%20the%20patch%27s%20content..

[the remark this answers is not in the document]
Comment (xiaozhuai) [reply ports/libimobiledevice-glue/004_fix_api.patch:9]: For sure, if you have questions about certain patches, you can bring them up and I'll explain why.

**Comment by Adela0814** [comment]
Usage tested pass on `x64-windows` and `x64-windows-static`.

**Review by Adela0814 — DISMISSED** [review]

**Review by dan-shaw — DISMISSED** [review]

File: ports/ideviceinstaller/vcpkg.json [hunk 7]
```diff
@@ -1,21 +1,21 @@
...
   "description": "Manage apps of iOS devices",
   "homepage": "https://libimobiledevice.org/",
-  "license": "LGPL-2.1-only",
-  "supports": "windows & !arm64 & !xbox",
+  "license": "LGPL-2.0-or-later",
+  "supports": "!uwp & !android & !ios",
```
[the remark this answers is not in the document]
Comment (xiaozhuai) [reply ports/ideviceinstaller/vcpkg.json]: Done~

**Comment by BillyONeal** [comment]
Given that the existing ports using the fork only work on Windows, and true upstream only works on non-Windows, would it make sense to make these 'meta' ports that choose a fork or the true upstream based on platform?

(This mess of attributing stuff to a fork is why we added checking for this to our new port checklist...)

**Comment by xiaozhuai** [comment]
> would it make sense to make these 'meta' ports that choose a fork or the true upstream based on platform?

@BillyONeal 

IMO, it's no need to do so. 
The affected ports are all tool only (`idevicerestore`, `ideviceinstaller`, `usbmuxd`)
For sure, the original upstream would also support windows, with just a few patches.
It's just not well maintained on msvc.
That's also why I add lines in ci baseline, not add `!windows` in supports.

I will take a look at them after this pr got merged.

**Comment by dg0yt** [comment]
> would it make sense to make these 'meta' ports that choose a fork or the true upstream based on platform?

IMO this would not create much comfort for cross-platform use cases but increase the confusion about the actual versions.

**Comment by xiaozhuai** [comment]
After investigation, `ideviceinstaller` now can build and work on windows.
But the `idevicerestore` and `usbmuxd` support for MSVC was very poor in original upstream and required a lot of code rewriting (Not simply patch).

Perhaps we should mark them as `(!windows & !uwp & !android & !ios & !xbox) | mingw`.

@dg0yt @BillyONeal What's your opinion?

**Comment by dg0yt** [comment]
https://github.com/libimobiledevice-win32/usbmuxd and https://github.com/libimobiledevice-win32/idevicerestore haven't been updated since 2020. IMO it is more important to follow and promote upstream. 


> Perhaps we should mark them as `(!windows & !uwp & !android & !ios & !xbox) | mingw`.

`supports` in manifest is a good way to indicate the current state of upstream platform support.
Omit uwp and xbox, they are covered by windows.

**Comment by xiaozhuai** [comment]
@BillyONeal @dan-shaw @dg0yt 
I think this PR is stable now. Time to merge : )

**Comment by xiaozhuai** [comment]
@BillyONeal @dan-shaw @dg0yt 
One more thing.
The `getopt-win32` port use `libimobiledevice-win32/getopt`, which is also not right.
I decide to open another pr to fix it after this pr got merged.
But I didn't find where is the original upstram.

**Comment by BillyONeal** [comment]
The problem is that for all existing customers of these ports, these changes break them. The fork is suboptimal, absolutely. We would not add ports using the fork today. However, as long as there is no platform support intersection between upstream and the fork, they can both be in the curated registry, and I think we owe not breaking everyone using these bits.

However I'd really like other maintainers to weigh in here.

**Comment by xiaozhuai** [comment]
> However, as long as there is no platform support intersection between upstream and the fork, they can both be in the curated registry

You mean split `idevicerestore` and `usbmuxd` into `*-unix` and `*-win`?

**Comment by BillyONeal** [comment]
> You mean split `idevicerestore` and `usbmuxd` into `*-unix` and `*-win`?

I think so but before spending a bunch of time doing that let's see what other maintainers think tomorrow.

**Comment by BillyONeal** [comment]
@ras0219-msft @JavierMatosD @dan-shaw and I discussed this today. We believe the following:

1. The name that corresponds with upstream must be changed to point at that upstream. (Which this PR does)
2. Within the port that normally points to upstream, we do not want to fetch a fork with different, possibly out of date contents. (That is, if we use the fork, there must be a metaport that chooses)
3. We could relax (2) if the fork is 'morally indistinguishable' from the original and tracks very quickly with identical versioning etc. (That is not the case here)

Other observations:
* How large are the changes in the fork to work with Windows? Is it just .def files or something else?
* We looked at telemetry for these ports and they are installed extremely rarely. As a result we are less concerned about introducing breaking changes to those customers. For instance, this was installed .02% as frequently as the most frequently installed port.

Agreed upon outcome:
* We should give @qmfrederik (and any other contributors we can find if possible) ~a week to respond before proceeding. If they reply that they are heavily using the fork, we should add libimobiledevice-win32 which is morally identical to the old ports for them to use. Note that the set intersection of the `-win32` ports and the real upstream ports is empty so they can both be in the registry.
* Otherwise, what's in this PR is good to go. ✅

=====================

Other things we discussed and discarded:
<details>

* Change these ports (for example `libimobiledevice`) to point at the real upstream, but when a platform upstream can't support is used, they become empty and depend on forked ports (for example `libimobiledevice-win32`)
* Do exactly what's in this pr, but add libimobiledevice-win32 which points at the forks. There's no metaport here, so folks who want the status quo have to change their declared dependencies. The supports expression intersections will be the empty set, so they can both be in the curated registry.
* Do exactly what's in this pr, but rename to libmobibledevice-original (name to be bikeshed) and deindex the original names currently used by the fork.

</details>

**Comment by qmfrederik** [comment]
This is the right thing to do.  My oldest patches for adding Visual Studio support to libimobiledevice date back to 2015 (https://github.com/libimobiledevice/libplist/pull/76, https://github.com/libimobiledevice/libusbmuxd/pull/28, https://github.com/libimobiledevice/libimobiledevice/pull/196), when CMake was relatively new, Visual Studio had no editor support for it, and the clang front-end for the msvc compiler wasn't a thing yet.

I've been wanting to redo libimobiledevice-win32 as a CMake project for a long time, and it's great to that someone else finally got round to doing this. Thanks, @xiaozhuai!

Getting usbmuxd to work on Windows is non-trivial; the only way I found to make it work is to use libusb-win32 and use a libusb driver for your iOS device.  In virtually all cases, it's easier to use the Apple Mobile Device service instead.

**Comment by BillyONeal** [comment]
@qmfrederik Thanks for confirming!

**Comment by BillyONeal** [comment]
@xiaozhuai Thanks for the PR!

**Pull request merged**